src/Nest/XPack/Security/Privileges/HasPrivileges/HasPrivilegesResponse.cs

```diff
@@ -0,0 +1,90 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Nest
+{
+	public interface IHasPrivilegesResponse : IResponse
+	{
+		[JsonProperty("username")]
+		string Username { get; }
+
+		[JsonProperty("has_all_requested")]
+		bool HasAllRequested { get; }
+
+		[JsonProperty("cluster")]
+		IReadOnlyDictionary<string, bool> Clusters { get; }
+
+		[JsonProperty("index")]
+		[JsonConverter(typeof(IndicesPrivilegesResponseJsonConverter))]
+		IReadOnlyCollection<ResourcePrivileges> Indices { get; }
+
+		[JsonProperty("application")]
+		[JsonConverter(typeof(ApplicationsPrivilegesResponseJsonConverter))]
+		IReadOnlyDictionary<string, IReadOnlyCollection<ResourcePrivileges>> Applications { get; }
+	}
+
+	public class HasPrivilegesResponse : ResponseBase, IHasPrivilegesResponse
+	{
+		public string Username { get; internal set;  }
+		public bool HasAllRequested { get; internal set; }
+
+		public IReadOnlyDictionary<string, bool> Clusters { get; internal set; } = EmptyReadOnly<string, bool>.Dictionary;
+
+		public IReadOnlyCollection<ResourcePrivileges> Indices { get; internal set; } = EmptyReadOnly<ResourcePrivileges>.Collection;
+
+		public IReadOnlyDictionary<string, IReadOnlyCollection<ResourcePrivileges>> Applications { get; internal set; } =
+			EmptyReadOnly<string, IReadOnlyCollection<ResourcePrivileges>>.Dictionary;
+	}
+
+	public class ResourcePrivileges
+	{
+		public string Resource { get; internal set; }
+
+		public IReadOnlyDictionary<string, bool> Privileges { get; internal set; } = EmptyReadOnly<string, bool>.Dictionary;
+	}
+
+	internal class IndicesPrivilegesResponseJsonConverter : JsonConverter
+	{
+		public override bool CanConvert(Type objectType) => true;
+		public override void WriteJson(JsonWriter writer, object value, JsonSerializer serializer) => throw new NotSupportedException();
+		public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+		{
+			var jsonObject = JObject.Load(reader);
```

Comment:
Not personally keen on using `JObject` here for new methods. I think a `while()` loop with Token type check will be faster and probably cleaner.

Reply:
I've added a while loop implementation